tls/s2n_x509_validator.c

```diff
@@ -0,0 +1,430 @@
+    if (store->trust_store) {
+        X509_STORE_free(store->trust_store);
+        store->trust_store = NULL;
+    }
+}
+
+static uint8_t default_verify_host(const char *host_name, size_t len, void *data) {
+    return 1;
+}
+
+int s2n_x509_validator_init_no_checks(struct s2n_x509_validator *validator) {
+    validator->trust_store = NULL;
+    validator->verify_host_fn = NULL;
+    validator->validation_ctx = NULL;
+    validator->cert_chain = NULL;
+    validator->validate_certificates = 0;
+    validator->check_stapled_ocsp = 0;
+
+    return 0;
+}
+
+int s2n_x509_validator_init(struct s2n_x509_validator *validator, struct s2n_x509_trust_store *trust_store,
+                            uint8_t check_ocsp, verify_host verify_host_fn, void *verify_ctx) {
+    notnull_check(trust_store);
+    validator->trust_store = trust_store;
+
+    validator->validate_certificates = 1;
+    validator->check_stapled_ocsp = check_ocsp;
+
+    validator->verify_host_fn = default_verify_host;
+
+    if (verify_host_fn) {
+        validator->verify_host_fn = verify_host_fn;
+    }
+
+    validator->validation_ctx = verify_ctx;
+
+    validator->cert_chain = NULL;
+    if (validator->trust_store->trust_store) {
+        validator->cert_chain = sk_X509_new_null();
+    }
+
+    return 0;
+}
+
+void s2n_x509_validator_cleanup(struct s2n_x509_validator *validator) {
+    if (validator->cert_chain) {
+        X509 *cert = NULL;
+        while ((cert = sk_X509_pop(validator->cert_chain))) {
+            X509_free(cert);
+        }
+
+        sk_X509_free(validator->cert_chain);
+    }
+
+    validator->trust_store = NULL;
+    validator->verify_host_fn = NULL;
+    validator->validation_ctx = NULL;
+    validator->validate_certificates = 0;
+}
+
+//for each name in the cert. Iterate them. Call the callback. If one returns true, then consider it validated,
+//if none of them return true, the cert is considered invalid.
+static uint8_t verify_host_information(struct s2n_x509_validator *validator, X509 *public_cert) {
+    uint8_t verified = 0;
+
+    //get each subject alternative name.
+    STACK_OF(GENERAL_NAME) *names_list = X509_get_ext_d2i(public_cert, NID_subject_alt_name, NULL, NULL);
+    GENERAL_NAME *current_name = NULL;
+    while (!verified && names_list && (current_name = sk_GENERAL_NAME_pop(names_list))) {
+        const char *name = (const char *) M_ASN1_STRING_data(current_name->d.ia5);
+        size_t name_len = (size_t) M_ASN1_STRING_length(current_name->d.ia5);
+
+        verified = validator->verify_host_fn(name, name_len, validator->validation_ctx);
+    }
+
+    GENERAL_NAMES_free(names_list);
+
+    //if none of those were valid, go to the subject name.
+    if (!verified) {
+        X509_NAME *subject_name = X509_get_subject_name(public_cert);
+        if (subject_name) {
+            int j = 0, i = -1;
+            while ((j = X509_NAME_get_index_by_NID(subject_name, NID_commonName, i)) >= 0) {
```

Comment:
This while loop is iterating through all the CommonNames in the Subject Name so that only the final CommonName is checked (which I believe matches the behaviour of NSS). Is that what we want? Should we enforce that each Cert have only one Common Name?

We can open an Issue for longer discussion if this will block the PR, but I wanted to make sure that this is known. 

See Section 4.2 of the [PKI Layer Cake](https://www.esat.kuleuven.be/cosic/publications/article-1432.pdf) paper.

Reply:
I don't think we should "enforce" every cert having a Common Name (especially in client mode), since there's no telling what we'd break.

It's a fair question to ask which one we want to rely on though (the root or the last intermediate). I'm not entirely sure which it should be.